crates/shared/src/sources/balancer_v2/pools/common.rs

```diff
@@ -77,6 +94,59 @@ impl PoolInfoFetching for PoolInfoFetcher {
             block_created,
         })
     }
+
+    fn fetch_common_pool_state(
+        &self,
+        pool: &PoolInfo,
+        batch: &mut Web3CallBatch,
+        block: BlockId,
+    ) -> BoxFuture<'static, Result<PoolState>> {
+        let pool_contract = self.base_pool_at(pool.address);
+        let paused = pool_contract
+            .get_paused_state()
+            .block(block)
+            .batch_call(batch);
+        let swap_fee = pool_contract
+            .get_swap_fee_percentage()
+            .block(block)
+            .batch_call(batch);
+        let balances = self
+            .vault
+            .get_pool_tokens(Bytes(pool.id.0))
+            .block(block)
+            .batch_call(batch);
+
+        // Because of a `mockall` limitation, we **need** the future returned
+        // here to be `'static`. This requires us to clone and move `pool` into
+        // the async closure - otherwise it would only live for as long as
+        // `pool`, i.e. `'_`.
+        let pool = pool.clone();
+        async move {
+            let (paused, _, _) = paused.await?;
+            let swap_fee = Bfp::from_wei(swap_fee.await?);
+
+            let (token_addresses, balances, _) = balances.await?;
```

Comment:
Since the futures are instantiated concurrently beforehand, this is roughly as performant as joining them here isn't it (modulo that we could return an error early if the last one failed fast)?

Reply:
Because these are already resolved when the `batch_call` future resolves, there should be no performance difference whether we `join` or not.